dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
+    (if pub-root (dart-add-to-analysis-roots (expand-file-name pub-root))
+      (dart-add-to-analysis-roots (expand-file-name current-dir)))))
+
+(defun dart-add-to-analysis-roots (dir)
+  "Add DIR to the analysis roots."
+  (add-to-list 'dart-analysis-roots dir)
+  (dart-set-analysis-roots))
+
+(defun dart-set-analysis-roots ()
+  "Send the current list of analysis roots to the analysis server."
+  (dart-analysis-server-send
+   "analysis.setAnalysisRoots"
+   `(("included" . ,dart-analysis-roots)
+     ("excluded" . nil))))
+
+(defun dart-analysis-server-send (method &optional params callback)
+  "Send the METHOD request to the server with optional PARAMS.
+
+If you provide a CALLBACK, it will be called with the JSON decoded
+response.  Otherwise, the output will just be checked."
+  (let ((req-without-id (dart-analysis-server-make-request method params)))
+    (dart-analysis-server-enqueue req-without-id callback)))
+
+(defun dart-analysis-server-make-request (method &optional params)
+  "Construct a request for the analysis server.
+
+The constructed request will call METHOD with optional PARAMS."
+  `((method . ,method) (params . ,params)))
+
+(defun dart-analysis-server-enqueue (req-without-id callback)
+  "Send REQ-WITHOUT-ID to the analysis server, call CALLBACK with the result."
+  (setq dart-analysis-server-next-id (1+ dart-analysis-server-next-id))
+  (let ((request
+         (json-encode (push `(id . ,(format "%s" dart-analysis-server-next-id))
+                            req-without-id))))
+    (dart-log (concat "Sent:\n" request))
+    (if callback
+        (push `(,dart-analysis-server-next-id . ,callback)
+              dart-analysis-server-callbacks)
+      (push
+       `(,dart-analysis-server-next-id . dart-analysis-server-no-error-callback)
+       dart-analysis-server-callbacks))
+    (process-send-string (dart-analysis-server-process dart-analysis-server)
+                         (concat request "\n"))))
+
+(defun dart-analysis-server-no-error-callback (response)
+  "If RESPONSE has an error, report it."
+  (-when-let (resp-err (assoc 'error response))
+    (dart-log (format "Response from server had error: %s" resp-err))))
+
+(defun dart-analysis-server-handle-response (callback response)
+  "Call CALLBACK with the parsed JSON RESPONSE from the analysis server."
+  (dart-log (concat "Received:\n" (format "%s" response)))
+  (funcall callback response))
+
+(defun dart-analysis-server-process-filter (das string)
+  "Handle the event or method response from the dart analysis server.
+
+The server DAS has STRING added to the buffer associated with it.
+Method responses are paired according to their pending request and
+the callback for that request is given the json decoded response."
+  (let ((buf (dart-analysis-server-buffer das)))
+    (when (buffer-live-p buf)
+      (with-current-buffer buf
+        (goto-char (point-max))
+        (insert string)
+        (let ((buf-lines (split-string (buffer-string) "\n")))
```

Comment:
Add a comment explaining why you're writing to and reading from a buffer here rather than just using the string itself.